executor/slow_query.go

```diff
@@ -447,14 +447,25 @@ func (e *slowQueryRetriever) parseSlowLog(ctx context.Context, sctx sessionctx.C
 		if e.stats != nil {
 			e.stats.readFile += time.Since(startTime)
 		}
+		failpoint.Inject("mockReadSlowLogSlow", func(val failpoint.Value) {
+			if val.(bool) {
+				signals := ctx.Value("signals").([]chan int)
+				signals[0] <- 1
+				<-signals[1]
+			}
+		})
 		for i := range logs {
 			log := logs[i]
 			t := slowLogTask{}
 			t.resultCh = make(chan parsedSlowLog, 1)
 			start := offset
-			wg.Add(1)
 			ch <- 1
-			e.taskList <- t
+			select {
+			case <-ctx.Done():
+				return
+			case e.taskList <- t:
+			}
+			wg.Add(1)
```

Comment:
```wg``` can use ```util.WaitGroupWrapper```. so it can simple this code.

```
wg.Run(func {
      result, err := e.parseLog(ctx, sctx, log, start)
      ....
})
```

it is never forgotten to write ```wg.Add``` and ```wg.Done```.